File: manual/code/generate_strategies/RA(method)/cpp/function_Whoosh_index_writer.cpp
#include <string>
#include <xapian.h>
#include <filesystem>
#include <sstream>

class WhooshIndexManager {
public:
    template<typename T>
    std::string manage_and_search_documents(const std::string& index_dir, 
                                          const T& document_text = T(), 
                                          const T& search_query = T()) {
        try {
            if (!std::filesystem::exists(index_dir)) {
                std::filesystem::create_directories(index_dir);
            }

            if constexpr (std::is_same_v<T, std::string>) {
                if (!document_text.empty()) {
                    Xapian::WritableDatabase db(index_dir, Xapian::DB_CREATE_OR_OPEN);
                    Xapian::Document doc;
                    doc.set_data(document_text);

                    Xapian::TermGenerator indexer;
                    indexer.set_document(doc);
                    indexer.set_stemmer(Xapian::Stem("english"));
                    indexer.index_text(document_text);

                    db.add_document(doc);
                    db.commit();
                }

                if (!search_query.empty()) {
                    Xapian::Database db(index_dir);
                    Xapian::QueryParser qp;
                    qp.set_database(db);
                    qp.set_stemmer(Xapian::Stem("english"));
                    qp.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);

                    Xapian::Query query = qp.parse_query(search_query);
                    Xapian::Enquire enquire(db);
                    enquire.set_query(query);

                    Xapian::MSet results = enquire.get_mset(0, 10);
                    if (results.empty()) {
                        return "No documents found";
                    }

                    std::ostringstream oss;
                    for (auto it = results.begin(); it != results.end(); ++it) {
                        if (it != results.begin()) {
                            oss << " ";
                        }
                        oss << it.get_document().get_data();
                    }
                    return oss.str();
                }
                return "No documents found";
            } else {
                return "No documents found";
            }
        } catch (const std::exception& e) {
            return std::string("Error: ") + e.what();
        }
    }
};

File: manual/code/generate_strategies/RA(method)/python/function_Whoosh_index_writer.py
from whoosh.index import create_in, open_dir
from whoosh.fields import Schema, TEXT
from whoosh.qparser import QueryParser
import os

class WhooshIndexManager:
    def manage_and_search_documents(self, index_dir: str, document_text: str = None, search_query: str = None) -> str:
        schema = Schema(content=TEXT(stored=True))
        
        if not os.path.exists(index_dir):
            try:
                os.makedirs(index_dir)
                create_in(index_dir, schema)
            except Exception as e:
                raise ValueError(f"Failed to initialize index directory: {index_dir}") from e

        ix = open_dir(index_dir)

        try:
            writer = ix.writer()
            if document_text:
                writer.add_document(content=document_text)
            writer.commit()
        except Exception as e:
            if 'writer' in locals() and not writer.is_closed:
                writer.cancel()
            raise e

        if search_query:
            with ix.searcher() as searcher:
                query = QueryParser("content", ix.schema).parse(search_query)
                results = [hit["content"] for hit in searcher.search(query)]
                return " ".join(results) if results else "No documents found"
        return "No documents found"
